lib/stream_data.ex

```diff
@@ -2094,6 +2094,85 @@ defmodule StreamData do
     end)
   end
 
+  @doc """
+  Generates arbitrary (past and future) dates.
+
+  Generated values shrink towards `Date.utc_today/0`.
+  """
+  @spec date() :: t(Date.t())
+  def date do
+    date([])
+  end
+
+  @doc """
+  Generates dates according to the given `options` or `date_range`.
+
+  ## Options
+
+  * `:origin` - (`Date`) if present, generated values will shrink towards this date. Cannot be combined with `:min` or `:max`.
+
+  * `:min` - (`Date`) if present, only dates _after_ this date will be generated. Values will shrink towards this date.
+
+  * `:max` - (`Date`) if present, only dates _before_ this date will be generated. Values will shrink towards this date.
+
+  If both `:min` and `:max` are provided, dates between the two mentioned dates will be generated.
+  Values will shrink towards `:min`.
+
+  If no options are provided, will work just like `date/0`.
+
+  ## `Date.Range`
+
+  Alternatively a `Date.Range` can be given. This will generate dates in the given range,
+  and with the supplied `date_range.step`.
+
+  Values will shrink towards `date_range.first`.
+
+  ## Calendar support
+
+  This generator works with `Calendar.ISO` and any other calendar
+  which implements the callbacks
+  `c:Calendar.naive_datetime_to_iso_days/7` and `c:Calendar.naive_datetime_from_iso_days/2`.
+  """
+  @spec date(Date.Range.t() | keyword()) :: t(Date.t())
+  def date(options_or_date_range)
+
+  def date(date_range = %Date.Range{}) do
+    Date.to_gregorian_days(date_range.first)..Date.to_gregorian_days(date_range.last)//date_range.step
+    |> StreamData.integer()
+    |> StreamData.map(&Date.from_gregorian_days(&1))
+  end
+
+  def date(options) when is_list(options) do
+    min = Keyword.get(options, :min, nil)
+    max = Keyword.get(options, :max, nil)
+    origin = Keyword.get(options, :origin)
+
+    case {min, max, origin} do
+      {nil, nil, nil} ->
+        StreamData.integer() |> StreamData.map(&Date.add(Date.utc_today(), &1))
+
+      {nil, nil, origin = %Date{}} ->
+        StreamData.integer() |> StreamData.map(&Date.add(origin, &1))
+
+      {nil, max = %Date{}, nil} ->
+        StreamData.positive_integer() |> StreamData.map(&Date.add(max, -&1))
+
+      {min = %Date{}, nil, nil} ->
+        StreamData.positive_integer() |> StreamData.map(&Date.add(min, &1))
+
+      {min = %Date{}, max = %Date{}, nil} ->
+        if Date.before?(max, min) do
+          raise ArgumentError, "max must not be before min"
```

Comment:
Let's show the given options.


```suggestion
          raise ArgumentError, """
          expected :max to be after or equal to :min, got:
          
            * min: #{inspect(min)}
            * max: #{inspect(max)}
          """
```